hbase-endpoint/src/test/java/org/apache/hadoop/hbase/coprocessor/TestAggregateImplementation.java

```diff
@@ -0,0 +1,958 @@
+      .setClientSupportsPartialResult(true).build();
+
+    when(env.checkScanQuota(any(), anyLong(), anyLong()))
+      .thenThrow(new RpcThrottlingException(ReadSizeExceeded, 1000, "Throttled for testing"));
+
+    ArgumentCaptor<AggregateResponse> responseCaptor =
+      ArgumentCaptor.forClass(AggregateResponse.class);
+    RpcCallback<AggregateResponse> callback = mock(RpcCallback.class);
+
+    // call gets no results, and response should contain the same start row as the request, because
+    // no progress in the scan was made
+    aggregate.getMax(controller, request, callback);
+
+    verify(callback).run(responseCaptor.capture());
+
+    AggregateResponse response = responseCaptor.getValue();
+
+    assertTrue("Response should indicate there are more rows", response.hasNextChunkStartRow());
+    assertEquals("response should contain the same start row as the request",
+      request.getScan().getStartRow(), response.getNextChunkStartRow());
+  }
+
+  @Test
+  public void testMaxWithNoResults() throws Exception {
+    AggregateRequest request = AggregateRequest.newBuilder().setScan(ProtobufUtil.toScan(scan))
+      .setInterpreterClassName(LongColumnInterpreter.class.getName())
+      .setClientSupportsPartialResult(false).build();
+
+    doAnswer(invocation -> false).when(scanner).next(any(List.class));
+
+    ArgumentCaptor<AggregateResponse> responseCaptor =
+      ArgumentCaptor.forClass(AggregateResponse.class);
+    RpcCallback<AggregateResponse> callback = mock(RpcCallback.class);
+
+    aggregate.getMax(controller, request, callback);
+    verify(callback).run(responseCaptor.capture());
+    AggregateResponse response = responseCaptor.getValue();
+    assertNull(response);
+  }
+
+  @Test
+  public void testMaxDoesNotSupportPartialResults() throws Exception {
+    AggregateRequest noPartialRequest =
+      AggregateRequest.newBuilder().setScan(ProtobufUtil.toScan(scan))
+        .setInterpreterClassName(LongColumnInterpreter.class.getName())
+        .setClientSupportsPartialResult(false).build();
+
+    OperationQuota nonThrottlingQuota = mock(OperationQuota.class);
+    when(env.checkScanQuota(any(), anyLong(), anyLong())).thenReturn(nonThrottlingQuota);
+
+    ArgumentCaptor<AggregateResponse> responseCaptor =
+      ArgumentCaptor.forClass(AggregateResponse.class);
+    RpcCallback<AggregateResponse> callback = mock(RpcCallback.class);
+
+    // Call should complete without throttling
+    aggregate.getMax(controller, noPartialRequest, callback);
+
+    verify(callback).run(responseCaptor.capture());
+
+    AggregateResponse response = responseCaptor.getValue();
+
+    assertFalse("Response should not indicate there are more rows",
+      response.hasNextChunkStartRow());
+    ByteString b = response.getFirstPart(0);
+    HBaseProtos.LongMsg q = getParsedGenericInstance(LONG_COLUMN_INTERPRETER.getClass(), 3, b);
+    assertEquals(5L, (long) LONG_COLUMN_INTERPRETER.getCellValueFromProto(q));
+  }
+
```

Comment:
Do you have coverage of the restored old behavior? Something like `max` when partial results are not supported and a throttle is triggered? Or is that covered by another test?

Reply:
I have coverage of all supported combinations of behavior. `partial results are not supported and a throttle is triggered` is not a supported combination -- when partial results are not supported, the coprocessor charges ahead without checking quotas.

Reply:
👍 

Thanks for clarifying.